Badges

sean-duffy opened the pull request.

Commit:
Defines db schema for badges system

Defines tables for user, badges, videos_watched and quiz_results and the relationships between them for tracking a user's progression.

Commit:
Submit quiz results to the database if the user is logged in

Commit:
Merge static string Id changes into Module.json from lecturer email

Commit:
Add simple example badges to the json

Commit:
Add Helpers/BadgeUtils.rb

This file will hold helper functions for dealing with badges - currently, it loads them all and gives them a numerical ID similarly to how the phases are loaded.
This helper has been included in main.rb, allowing it's functions to be called as if they were declared in that file.

Commit:
Add phase_utils to helpers.

This helper splits out the get_by_id functionality and the work to load phases from main.rb,
reducing main.rb's complexity and improving clarity.

Commit:
Remove duplication with phase_utils helper.

Main instead will now use the PhaseUtils helper for get_by_id and loading of phase content.

Commit:
remove commented code

Commit:
Change json_id from type Integer to String

This reflects the decision to user string identifiers in the module.json

Commit:
Add helper methods to badge_utils
Adds methods to check if a page should trigger any badges
and to check the requirements to see if a badge should be rewarded

Commit:
prefix badgeutils in main due to broken commit

Commit:
Replace video["id"] with numeric value
Returns to prevoious behaviour, overwrriting the string ids.

This will need to be altered in the future to allow badges to trigger!

Commit:
Differentiate between String ID in JSON and calculated position id

Previously all resources were identified in order of appearance,
but now we also have a string identifier defined in the json.

This changes behaviour to not overwrite the string ID, and to differentiate
between that id and the positional id, such that flippd behaves as it did before (i.e.
next, prev links can be found by getting the prev and next page by order of apperarance)

Commit:
Add string identifiers to the example badges

Commit:
Add guards to requirement checks and change create_badge to
award_badge, which operates on a badge json object

Commit:
Tidy up badge helper for clarity
Add badge check/trigger to quiz post for quiz badges

GarethH9 committed:
Added video watched checkbox and AJAX POST request on video.erb page, and added logic to handle POST request and mark videos as watched and assign badges in main.rb

GarethH9 committed:
Added placeholder notification message (javascript alert) in videos.erb

sean-duffy committed:
Update build status to use correct branch

Commit:
Hide the tickbox for watching the video if you ar not signed in

Commit:
Factored out post methods from main into post and standardised indentation in main

Commit:
Resolved conflict

GarethH9 committed:
Added AJAX notifications which are displayed when a badge is earned from the videos.erb page

GarethH9 committed:
Notification is now displayed on awarding a badge for quiz results

Commit:
Add page showing a user's earnt badges, accessible under the
user dropdown on the right of the navbar.

Commit:
Remove duplication in main.rb / post.rb caused by merging upstream

Commit:
Remove unneeded before do in routes/badges,
it can access everything from main's before do

sean-duffy commented on app/routes/main.rb:
```diff
@@ -74,66 +40,68 @@ def get_by_id(phases, id)
-
-    erb :quiz_result
+	
+	get '/videos/:pos' do
```
What happened to the indentation here?

tylrbvn replied:
@sean-duffy  Looks like Katie's indentation standardisation didn't quite go to plan haha dc71267

Commit:
Corrected more indentation in main and post

Commit:
Factored out get_user_id and display_notification into GeneralUtils

Commit:
Factored quiz and video methods from main and post into quizes and videos files

Commit:
Added files referenced in previous commit

Commit:
Removed now unused post.rb

Commit:
Refactored videos.rb for increased clarity and reduced duplication

Commit:
Added tracking of general_utils

Commit:
temp commit - broken

Commit:
Fixed error in previous commit, factored out previous and next page link definition into utility method

Commit:
Added comments and removed unused helpers

Commit:
Add lecturer field to module.json to define who is a lecturer
Lecturer users can see the count of badges awarded to students on the my badges page

Commit:
Redirect the user if they log out on the badge page
As they should not be able to see badge things unless logged in

Commit:
Merge branch 'badges' of https://github.com/sean-duffy/flippd into badges

Conflicts:
	app/routes/init.rb

Commit:
fix merge conflict

GarethH9 committed:
Notifications now correclty display over other content on the page (fixed issue with z-index of notifications preventing anything at the top of the page being clickable)

GarethH9 committed:
Added images for badges to the badges.erb page and updated module.JSON to link to this location

kcmills committed:
Added tests for first two general utils, corrected if statement in get_user_id

kcmills committed:
Merge branch 'badges' of github.com:sean-duffy/flippd into badges

GarethH9 committed:
Added images for existing badges and updated these in module.JSON

GarethH9 committed:
Submitting a quiz with no answers no longer produces an error and gives the user a score of 0

kcmills committed:
Merged changes

kcmills committed:
Fixed test for if no quiz answers are selected

GarethH9 committed:
Removed <a> tag from badges on badge summary page as these don't currently link anywhere

GarethH9 committed:
Removed excessive padding from notifications to ensure that you can click all elements visible beneath the notification

GarethH9 committed:
Fixed 'Plug-ins' YouTube video in module.JSON

GarethH9 committed:
Increased field size for 'json_id' as some IDs in module.json were longer than the default limit. When this happened those records weren't being added to the database

Commit:
Moved notification margin to on top of the alert so you can now click links that are just below the bottom notification. (Note: You still won't be able to click links above or between notifications, but there isn't really anything that can be done about this)
